### ISSUE TITLE
Docker Performance Slow Compared to Binary

### ISSUE THREAD

**Issue by nickzelei** [vektra/mockery#822]
Description
-------------

I'm trying to understand more about why running mockery seems to be so much slower than when run bare metal.

I read through the issue todo and cleaned out my go mod cache.

Mockery Version
--------

v2.46.1

Go Version
--------------

go version go1.23.1 darwin/arm64

Installation Method
-------------------

- [ ] Binary Distribution
- [x] Docker
- [x] brew
- [ ] go install
- [ ] Other: [specify]

Steps to Reproduce
------------------
You can repro this in the [neosync](https://github.com/nucleuscloud/neosync) repo.
1. `go clean -modcache`
2. `go mod download`
3. `mockery` ~ 3 seconds
5. `docker run -v "$PWD":/src -w /src vektra/mockery:v2.46.1` -- ~1 minute

I've even tried mounting only what is necessary via:
```
docker run \
  -v "$PWD/backend":/src/backend \
  -v "$PWD/worker":/src/worker \
  -v "$PWD/internal":/src/internal \
  -v "$PWD/.mockery.yml":/src/.mockery.yml \
  -v "$PWD/go.mod":/src/go.mod \
  -v "$PWD/go.sum":/src/go.sum \
  -v "${GOPATH:-$HOME/go}/pkg/mod":/go/pkg/mod \
  -w /src \
  vektra/mockery:v2.46.01
```
But the docker run still takes orders of magnitude longer than the bare metal.

Am I missing some form of cache that should be mounted into the docker container?

### Expected Behavior

The docker run should be near the speed of the brew installed version.

### Actual Behavior

Takes much longer.

**Comment by LandonTClipp** [comment]
This is going to require some benchmarking. In a case like this, I'd try a few things:

1. Run mockery with `strace -T`in both docker and non-docker contexts to see how long it spends in syscalls.
2. Run mockery with `--log-level=DEBUG` to see if there's an obvious step where it spends too much time.

Does mockery run slow even after multiple iterations in docker (I wonder if this is a cache issue)? What version of docker are you running?

**Comment by rezikovka** [comment]
Same problem. 
Docker Desktop 4.34.3 (170107) is currently the newest version available.
In my project brew-installed mockery works in ~7 sec, but docker container takes several minutes

**Comment by LandonTClipp** [comment]
Sadly, this might be expected. Docker on Macs must run inside of a VM, which is going to incur a noticeable performance penalty. This is what Docker Desktop does under the hood. Although, an order of magnitude is _not_ exactly expected. This might be a factor of how the VM is configured. I cannot debug environment-specific issues like this so I will have to rely on others to identify where mockery is spending most of its time. I bet it is IO related.

**Comment by RangelReale** [comment]
I also have this feeling that when running in a `Dockerfile`, mockery is a lot slower than "native", in my M2 running directly it takes about 4 seconds, but running on docker build using the same architecture (`--platform linux/arm64`), it takes about 55 seconds, in our k8s CI jobs it takes 2 minutes.

If course running in docker it is expected to be slower, but this seems too much, even the go building don't have this difference in time.

I'm seeing if I can discover where this time difference happens, not very easy to do profiling in docker images.

**Comment by sr9000** [comment]
@nickzelei it is very likely that you forgot to `go mod download` inside your docker image before run mockery

in my case mockery cannot find private packages (remove `--quiet` option to see import errors) that makes it more than 10 times slower (order of magnitude).

P.S.
Also do not forget to using native docker images for your host arch. I my case https://hub.docker.com/_/golang does everything right.

**Comment by nickzelei** [comment]
> [@nickzelei](https://github.com/nickzelei) it is very likely that you forgot to `go mod download` inside your docker image before run mockery
> 
> in my case mockery cannot find private packages (remove `--quiet` option to see import errors) that makes it more than 10 times slower (order of magnitude).
> 
> P.S. Also do not forget to using native docker images for your host arch. I my case https://hub.docker.com/_/golang does everything right.

I don't believe that is the case. If you see in my original example I am mounting my go mod directory, so there shouldn't need to be any difference there.

I run a variety of tools inside of docker and the performance is negligible. The only one that I've seen a slowdown on mockery.

**Comment by RangelReale** [comment]
> > [@nickzelei](https://github.com/nickzelei) it is very likely that you forgot to `go mod download` inside your docker image before run mockery
> > in my case mockery cannot find private packages (remove `--quiet` option to see import errors) that makes it more than 10 times slower (order of magnitude).
> > P.S. Also do not forget to using native docker images for your host arch. I my case https://hub.docker.com/_/golang does everything right.
> 
> I don't believe that is the case. If you see in my original example I am mounting my go mod directory, so there shouldn't need to be any difference there.
> 
> I run a variety of tools inside of docker and the performance is negligible. The only one that I've seen a slowdown on is mockery.

Yes I also believe this is not the case, I run `go mod download` before running mockery.
Something is happening only in mockery for it to be so slow in docker, the performance difference is too big compared to other things (like even the Go compiler).

**Comment by frankjkelly** [comment]
Seeing this also.
When I add `--log-level=DEBUG` the slowness appears to be at the package download time.

```
17 Mar 25 15:22 UTC DBG getting packages section dry-run=false package-path=github.com/apache/pulsar-client-go/pulsar parent-package=github.com/apache/pulsar-client-go/pulsar sub-package=github.com/apache/pulsar-client-go/pulsar/log version=v2.53.2
17 Mar 25 15:23 UTC DBG found file dry-run=false file=/src/vendor/github.com/apache/pulsar-client-go/pulsar/auth/athenz.go package=github.com/apache/pulsar-client-go/pulsar/auth version=v2.53.2
```

**Comment by LandonTClipp** [comment]
That log line you posted doesn't suggest it's package downloading that's the issue. The line mentions that it's grabbing a particular config section from the in-memory config.

Could you confirm the CPU arch of your host and the arch of the image you downloaded?

**Comment by frankjkelly** [comment]
@LandonTClipp Thanks for the response - I get the following

```
docker run --security-opt label=disable -v .:/src -w /src docker.io/vektra/mockery:v2.53.2 --profile
Error: unknown flag: --profile
Usage:
  mockery [flags]
  mockery [command]
```

**Comment by frankjkelly** [comment]
I figured it must be `--cpuprofile` and here's what I have  (obviously it's not a text file)

[mockery.txt](https://github.com/user-attachments/files/19292595/mockery.txt)

My Architecture
```
$ uname
Darwin
$uname -m
arm64
```

**Comment by frankjkelly** [comment]
```
$ docker exec -it <containerid> sh
/src # uname -a
Linux 50ffcbdca0bc 6.10.14-linuxkit #1 SMP Mon Feb 24 16:35:16 UTC 2025 x86_64 Linux
```

and
```
[
    {
        "Id": "sha256:3405605badc07ce01febadc9c61df1222a02784359021970b31657d253164852",
        "RepoTags": [
            "vektra/mockery:latest",
            "vektra/mockery:v2.53.2"
        ],
        "RepoDigests": [
            "vektra/mockery@sha256:3405605badc07ce01febadc9c61df1222a02784359021970b31657d253164852"
        ],
        "Parent": "",
        "Comment": "buildkit.dockerfile.v0",
        "Created": "2025-03-10T16:25:51.458792006Z",
        "DockerVersion": "",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GOLANG_VERSION=1.24.1",
                "GOTOOLCHAIN=local",
                "GOPATH=/go",
                "GOCACHE=/tmp/.cache"
            ],
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "/go",
            "Entrypoint": [
                "/usr/local/bin/mockery"
            ],
            "OnBuild": null,
            "Labels": {
                "org.opencontainers.image.created": "2025-03-10T16:23:47Z",
                "org.opencontainers.image.name": "mockery",
                "org.opencontainers.image.revision": "bfd46e35b15c2689ced221299bdcdeeff8aa0be3",
                "org.opencontainers.image.source": "https://github.com/vektra/mockery",
                "org.opencontainers.image.version": "2.53.2"
            }
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 2529,
        "GraphDriver": {
            "Data": null,
            "Name": "overlayfs"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:08000c18d16dadf9553d747a58cf44023423a9ab010aab96cf263d2216b8b350",
                "sha256:040b911ff035de1c4531b814f51d4694785fdfbf6bdb07b1bc865d48b69de449",
                "sha256:e0f4cf0bdc482691782fcd0642612f0aa8ffc90ef2506f0619ab2507d726bafc",
                "sha256:4141acd56718d0d527f7376a0646f372a9d535b73ccd7d2a4dcb39b8e133539b",
                "sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef",
                "sha256:70686b0e2a02cbfc6528a6f7c60a4d644740cfbce03c47deb9b15ed6b25ea6ee",
                "sha256:f91ea88287ed1bc0179adad39800d98ceaf0b40e8b6102febb242228e7c6d40a"
            ]
        },
        "Metadata": {
            "LastTagTime": "2025-03-17T17:21:38.511569208Z"
        },
        "Descriptor": {
            "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
            "digest": "sha256:3405605badc07ce01febadc9c61df1222a02784359021970b31657d253164852",
            "size": 743
        }
    }
]

```
Thank you

**Comment by LandonTClipp** [comment]
Okay, here's a MAJOR smoking gun. Your host arch is arm64 and you've pulled down the amd64 image. That's going to cause a large amount of emulation overhead and probably the reason for your slowness.

Can you try running mockery with the `vektra/mockery:v2.53.2-arm64` tag?

**Comment by LandonTClipp** [comment]
Another thing to investigate is whether or not the VM that is running your docker containers is of the same architecture as your host MacOS (which you've shown to be amd64). If you're using Docker Desktop, I'm assuming it's smart enough to use the same arch. I use colima typically so I'm not sure how to verify that on Docker Desktop.

**Comment by frankjkelly** [comment]
@LandonTClipp I think you mean I should run an `arm64` tagged image?

I ran
```
docker run --security-opt label=disable -v .:/src -w /src docker.io/vektra/mockery:v2.53.2-arm64
```
and it took about 30 seconds (half the time)

I also tried 
```
docker run --security-opt label=disable -v .:/src -w /src docker.io/vektra/mockery:v2.53.2-amd6
```
and it was the same speed as the original about 90 seconds to 2 minutes.

**Comment by LandonTClipp** [comment]
Yeah I meant `arm64`. 30 seconds is still too long IMO (but obviously it depends on how many interfaces it's parsing). If you find that bare metal is still much faster than the arm64 image, you can do a [profile diff](https://github.com/google/pprof/blob/main/doc/README.md#comparing-profiles) to see where the differences lie. Although in the profile you gave me, I pretty much see across-the-board slowness, but especially in garbage collection.

Ultimately, this appears an environmental problem unless we can narrow down a specific reason why docker decided to download the wrong arch. I would suggest looking into your VM's architecture, or possibly trying to boot a VM through `colima`, logging into the VM, and testing performance of a dockerless run of mockery inside the VM and compare it to docker-in-VM. This will help narrow down whether the slowness comes from the VM or if docker itself is to blame.

Docker on MacOS is just kind of annoying due to the need to run this VM. I've encountered this in many different places.

![Image](https://github.com/user-attachments/assets/d681d6f2-fecc-4c4a-991c-7b4b8b717d65)

**Comment by frankjkelly** [comment]
@LandonTClipp Thanks - the proper image is tolerable enough for me. Thanks again!
